sqlglot/dialects/oracle.py

```diff
@@ -282,3 +282,180 @@ def add_column_sql(self, expression: exp.AlterTable) -> str:
+
+    Returns:
+        The updated dictionary of joins
+    """
+    if join.alias_or_name in join_dict:
+        join_dict[join.alias_or_name].set(
+            "on",
+            exp.and_(join_dict[join.alias_or_name].args["on"], join.args["on"]),
+        )
+    else:
+        join_dict[join.alias_or_name] = join
+    return join_dict
+
+
+def _clean_binary_node(node: exp.Expression) -> None:
+    """if the node is left with only one child, promote the child to the parent node.
+    transformation is done in place.
+
+    Args:
+        node: The node to clean"""
+    if isinstance(node, exp.Binary):
+        if node.left is None:
+            node.replace(node.right)
+        elif node.right is None:
+            node.replace(node.left)
+
+
+def _has_join_mark(col: exp.Expression) -> bool:
+    """Check if the column has a join mark
+
+    Args:
+        The column to check
+    """
+    if not isinstance(col, exp.Column):
+        return False
+    result = col.args.get("join_mark", False)
+    return result
+
+
+def _equality_to_join(
+    eq: exp.Binary, old_joins: t.Dict[str, exp.Join], old_from: exp.From
+) -> t.Optional[exp.Join]:
+    """Convert an equality predicate to a join if it contains a join mark
+
+    Args:
+        eq: The equality expression to convert to a join
+
+    Returns:
+        The join expression if the equality contains a join mark (otherwise None)
+    """
+    if not (isinstance(eq.left, exp.Column) or isinstance(eq.right, exp.Column)):
+        return None
+    new_eq = eq.copy()
+    left_has_join_mark = _has_join_mark(eq.left)
+    right_has_join_mark = _has_join_mark(eq.right)
+
+    if left_has_join_mark:
+        new_eq.left.set("join_mark", False)
+        join_on = new_eq.left.table
+    elif right_has_join_mark:
+        new_eq.right.set("join_mark", False)
+        join_on = new_eq.right.table
+    else:
+        return None
+
+    join_this = old_joins.get(join_on, old_from).this
+    return exp.Join(this=join_this, on=new_eq, kind="LEFT")
+
+
+def _eliminate_join_marks_from_scope(scope) -> None:
+    """Remove join marks columns in scope's where clause.
+    Converts them to left joins and replaces any existing joins.
+    Updates scope in place.
+
+    Args:
+        scope: The scope to remove join marks from
+    """
+    select_scope = scope.expression
+    if not select_scope.args.get("where"):
+        return
+    if not select_scope.args.get("joins"):
+        return
+
+    # dictionaries used to keep track of joins to be replaced
+    old_joins: t.Dict[str, exp.Join] = {
+        join.alias_or_name: join for join in list(select_scope.args.get("joins", []))
+    }
+    new_joins: t.Dict[str, exp.Join] = {}
+
+    for node in scope.find_all(exp.Column):
+        if _has_join_mark(node):
+            # find the predicate that contains the join mark
+            predicate = node.find_ancestor(exp.Predicate)
+            if not isinstance(predicate, exp.Binary):
+                continue
```

Comment:
Also noting this bit:

> If two tables are joined over multiple join conditions, you must use the (+) operator in all or none of these conditions. A join with mixed syntax styles runs as an inner join, without warning.

Which is actually part of Redshift's [docs](https://docs.aws.amazon.com/redshift/latest/dg/r_WHERE_oracle_outer.html). Redshift's supposed to mimic Oracle's semantics AFAICT, so I'd expect this rule to also apply to Oracle.

If I understand correctly, this means that we should also avoid converting the `(+)` syntax if there's at least one condition involving the joined tables that doesn't use it. Otherwise you'd be silently changing this "A join with mixed syntax styles runs as an inner join, without warning." into "A join with mixed syntax styles will be converted into an outer join whose conditions will contain the marked columns, and the rest of them will remain as a filter".

Reply:
Btw, Redshift supporting this operator also means that we should probably factor this logic out into `transforms.py` so that it can be reused by both it and Oracle.

Reply:
> This doesn't look correct. AFAICT, this `continue` being reachable implies that there's an execution path where we may do a partial elimination of the join markers, ending up with a query that combines both the marker syntax and the `FROM`-`JOIN` syntax. However, this is invalid Oracle SQL according to the [docs](https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/Joins.html#GUID-29A4584C-0741-4E6A-A89B-DCFAA222994A):
> 
> > You cannot specify the (+) operator in a query block that also contains FROM clause join syntax.
> 
> Perhaps we should instead take this assumption for granted, following the usual garbage in - garbage out philosophy.

This does make that assumption and continues processing rather than returning an error.

Reply:
we should be using the oracle docs: https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/Joins.html

There seems to be some contradiction with what you describe.

Reply:
Any conditions imposed by the docs should be included as a test case